src/lib/marshal/workflow/writer.ts

```diff
@@ -212,41 +217,29 @@ const compileExtractionSettings = (
   );
 };
 
-/*
- * For a given workflow payload (and its local workflow reference), this function
- * builds a "workflow directory bundle", which is an obj made up of all the
- * relative file paths (within the workflow directory) and its file content to
- * write the workflow directory.
- *
- * Every workflow will always have a workflow.json file, so every bundle includes
- * it and its content at minimum. To the extent the workflow includes any
- * extractable fields, those fields content get extracted out and added to the
- * bundle.
- *
- * Important things to keep in mind re: content extraction:
- * 1. There can be multiple places in workflow json where content extraction
- *    happens.
- * 2. There can be multiple levels of content extraction happening, currently
- *    at a maximum of 2 levels.
- *
- * The way this function works and handles the content extraction is by:
- * 1. Traversing the given step node, and compiling all annotated extraction
- *    settings by the object path in the node *ordered from leaf to root*.
- * 2. Iterate over compiled extraction settings from leaf to root, and start
- *    extracting out the field as needed. In case the node that needs to be
- *    extracted out contains extracted file paths, then those file paths get
- *    rebased to relative to the referenced file.
- */
-const buildWorkflowDirBundle = (
-  remoteWorkflow: WorkflowData<WithAnnotation>,
-  localWorkflow: AnyObj = {},
-): WorkflowDirBundle => {
-  const bundle: WorkflowDirBundle = {};
-  const mutWorkflow = cloneDeep(remoteWorkflow);
+const keyLocalWorkflowStepsByRef = (
+  steps: AnyObj[] = [],
```

Comment:
```suggestion
  steps: unknown,
```
I think we want to treat the local workflow and its data a bit more cautiously as unknown and then validate it as we go, rather than assuming a certain shape. 

So something like:
```
const keyLocalWorkflowStepsByRef = (
  steps: unknown = [],
  result: AnyObj = {},
) => {
  if (!Array.isArray(steps)) {
    return result
  }

  for (const step of steps) {
    if (!isPlainObject(step)) continue
    if (!step.ref) continue

    result[step.ref] = step;

    if (step.type === StepType.IfElse) {
      const branches = step.branches || []
      if (!Array.isArray(branches)) continue

      for (const branch of branches) {
        if (!isPlainObject(branch)) continue

        result = keyLocalWorkflowStepsByRef(branch.steps, result);
      }
    }
  }

  return result;
};
```